traitlets/traitlets.py

```diff
@@ -2689,3 +2602,101 @@ def info(self):
         if self.allow_none:
             return result + " or None"
         return result
+
+#-----------------------------------------------------------------------------
+# Casting TraitTypes
+#-----------------------------------------------------------------------------
+
+class CastingMixin(TraitType):
+    """A mixin class for TraitTypes which cast inputs
+
+    Attributes
+    ----------
+    cast_to : callable
+        The callable which will be casting inputs.
+    _cast_types : tuple of classes
+        An attribute that can be assigned in subclasses
+        to limit what types of values are allowed to be
+        cast (this attribute is not defined in the base
+        class for mixin purposes).
+    """
+
+    cast_to = None
+
+    def validate(self, obj, value):
+        cast_types = getattr(self, '_cast_types', All)
+        if cast_types is not All:
+            for cls in cast_types:
+                if isinstance(value, cls):
+                    break
+            else:
+                self.error(obj, value)
+        try:
+            value = self.cast_to(value)
+        except Exception as e:
+            self.error(obj, value)
+        s = super(CastingMixin, self)
+        if hasattr(s, "validate"):
+            value = s.validate(obj, value)
+        return value
+
+    def info(self):
+        cast_types = getattr(self, '_cast_types', All)
+        if cast_types is not All:
+            return " or ".join(class_of(c) for c in set(cast_types).union([self.cast_to]))
+        else:
+            return "cast to %s" % (self.info_text or class_of(self.cast_to))
```

Comment:
class_of isn't going to provide the right info for the cases where `cast` isn't a type (e.g. re.compile, below).